File: haskell-side/src/Chapter5.hs
module Chapter5 where

-- Model an RPG Character with Abilities, Attributes and Stats

newtype Strength = Strength Int deriving (Show)

newtype Dexterity = Dexterity Int deriving (Show)

newtype Wisdom = Wisdom Int deriving (Show)

newtype Charisma = Charisma Int deriving (Show)

newtype Constitution = Constitution Int deriving (Show)

newtype Intelligence = Intelligence Int deriving (Show)

data Attributes = Attributes Strength Dexterity Wisdom Charisma Constitution Intelligence deriving (Show)

newtype HitPoints = HitPoints Int deriving (Show)

data Character = Character Attributes HitPoints deriving (Show)

File: scala-side/src/main/scala/com/geophy/Chapter5.scala
package com.geophy

case class Dexterity(stat: Int)
case class Wisdom(stat: Int)
case class Charisma(stat: Int)
case class Constitution(stat: Int)
case class Intelligence(stat: Int)
case class Strength(stat: Int)

case class Attributes(strength: Strength, dexterity: Dexterity, wisdom: Wisdom, charisma: Charisma, constitution: Constitution, intelligence: Intelligence)

type MaxHitPoints = Int

case class Character(attributes: Attributes, maxHitPoints: MaxHitPoints)

enum Spell:
  case Fireball
  case EldritchBlast

type CurrentHitPoints = Int
type Damage = Int

object Chapter5 extends App:
  val myCh = Character(
    ???,
    100
  )
